django/db/models/base.py

```diff
@@ -1080,7 +1080,10 @@ def _save_table(
         if pk_val is None:
             pk_val = meta.pk.get_pk_value_on_save(self)
             setattr(self, meta.pk.attname, pk_val)
-        pk_set = pk_val is not None
```

Comment:
You mean to make this a class function? Yeah we could.

Reply:
Although, I'm not sure if this would be practical, we would need to add `is_set` to every field. Maybe it's simpler this way.

Reply:
Assuming there doesn't already exist a method on `Field` that could serve this purpose (it's possible there is) then simply adding one and overriding it on `CompositePrimaryKey` (which sublasses it) seem like a better approach given this will raise in 99% of all `save` calls (as most models won't use a composite primary key).

Reply:
You're right, this would be bad for performance. I'm not sure about extending the `Field` class just for this, how about we do a simple `hasattr` check instead?

Reply:
I'd be surprised if the performance difference mattered in practice. That said, I think adding the method to `Field` feels cleanest to me.

Reply:
Okay, let's go with the method. I agree it doesn't matter 99.99% of the time, but this might be an exception (no pun intended) because this code is so widely used

Reply:
After having reviewed PR #18450, I'm honestly confused about why we would put the "responsibility" of knowing about a PK value in a `Field` and not in the `Model` instance. More so, the new method as proposed does not use anything from the `Field` class (and I think it will not either in the `CompositePrimaryKey` override).

Could we do something like this for `ModelBase` (hand wavy untested code)?

* for `main`:
```python
    def is_pk_set(self):
        return self.pk is not None
```
* this branch:
```python
    def is_pk_set(self):
        if isinstance(self.pk, CompositePrimaryKey):
            return all(value is not None for value in self.pk)
        return self.pk is not None
```